File: 2020/13/two/main.go
package main

// this is broken!

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func readFile() []string {
	file, err := os.Open("input")
	if err != nil {
		log.Fatalf("Failed to open file.")
	}
	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)
	var text []string
	for scanner.Scan() {
		text = append(text, scanner.Text())
	}
	file.Close()
	return text
}

func run(starttime int, ids []int) {
	time := starttime
	step := ids[0]
	for i, id := range ids {
		if id < 0 && i > 0 {
			continue
		}
		j := 0
		for {
			j++
			fmt.Println(j)
			k := time + j*step
			if (k+i)%id == 0 {
				time = k
				step *= id
				break
			}
		}
		// fmt.Println(id)
	}
	fmt.Println(time)
}

func main() {
	lines := readFile()
	notes := lines[1]
	notesarr := strings.Split(notes, ",")

	ids := make([]int, 0)
	for _, note := range notesarr {
		if note != "x" {
			id, err := strconv.Atoi(string(note))
			if err != nil {
				log.Fatal(err)
			}
			ids = append(ids, id)
		} else if note == "x" {
			ids = append(ids, -1)
		}
	}

	run(0, ids)
}

File: 2020/13/two/main.lua
local lines = {}
for line in io.lines() do
    lines[#lines+1] = line
end

local notes = lines[2]

local indices = {}
local buses = {}

local k = 0
for bus in string.gmatch(notes, "%w+") do
    k = k + 1
    if bus == "x" then
    
    else
        buses[k] = bus
        if k ~= 0 then
            indices[#indices+1] = k
        end
    end
end

local time = 0
local step = 1
for _, i in ipairs(indices) do
    local j = 0
    print(i, buses[i])
    while true do
        j = j + 1
        local t = time + j * step
        local bus = buses[i]
        local test = (t+i-1)%bus
        if test == 0 then
            step = step * bus
            time = t
            break
        end
    end
end
print(string.format("%.0f",time ))